lab_2_tokenize_by_bpe/main.py

```diff
@@ -26,42 +38,88 @@ def collect_frequencies(
+        preprocessed_word = prepare_word(word, start_of_word, end_of_word)
+        if not preprocessed_word:
+            return None
+        freq_dict.update({preprocessed_word: words.count(word)})
+    return freq_dict
 
 
 def count_tokens_pairs(
-    word_frequencies: dict[tuple[str, ...], int]
+        word_frequencies: dict[tuple[str, ...], int]
 ) -> dict[tuple[str, str], int] | None:
     """
     Counts number of occurrences of each pair of subsequent tokens
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+    pairs_of_tokens = {}
+    for word, freq in word_frequencies.items():
+        for ind in range(len(word) - 1):
+            token_pair = (word[ind], word[ind + 1])
+            if token_pair not in pairs_of_tokens:
+                pairs_of_tokens[token_pair] = freq
+            pairs_of_tokens[token_pair] += freq
+    return pairs_of_tokens
 
 
 def merge_tokens(
-    word_frequencies: dict[tuple[str, ...], int], pair: tuple[str, str]
+        word_frequencies: dict[tuple[str, ...], int], pair: tuple[str, str]
 ) -> dict[tuple[str, ...], int] | None:
     """
     Updates word frequency dictionary by replacing a pair of token with a merged one
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :param pair: a pair of tokens to be merged
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(pair, tuple):
+        return None
+    merged_dict = {}
+    for word, freq in word_frequencies.items():
+        new_word = ''.join(word).replace(pair[0] + '' + pair[1], pair[0] + pair[1])
+        new_word_tupled = tuple(new_word.split())
+        if new_word_tupled not in merged_dict:
+            merged_dict[new_word_tupled] += freq
+        else:
+            merged_dict[new_word_tupled] = freq
+    return merged_dict
 
 
 def train(
-    word_frequencies: dict[tuple[str, ...], int] | None, num_merges: int
+        word_frequencies: dict[tuple[str, ...], int] | None, num_merges: int
 ) -> dict[tuple[str, ...], int] | None:
     """
     Creates required number of new tokens by merging existing ones
     :param word_frequencies: dictionary of a kind <preprocessed word: number of occurrences>
     :param num_merges: required number of new tokens
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(num_merges, int):
+        return None
+    while num_merges > 0:
+        pairs_of_tokens = count_tokens_pairs(word_frequencies)
+        if not pairs_of_tokens:
+            return None
+        if num_merges > len(pairs_of_tokens):
+            num_merges = len(pairs_of_tokens)
```

Comment:
Can you explain your logic here?

Reply:
If there are not enough token pairs to merge num_merges times, num_merges is set to the number of available pairs? But maybe it should be done the other way, I'll try to alter the code maybe...

Reply:
Like we can't do more merges than the quantity of elements in our dict, as far as I understood